### ISSUE TITLE
[WIP] Namespace support for IPFS

### ISSUE THREAD

**Issue by dellagustin** [Podcastindex-org/podcast-namespace#136]
I'm still researching IPFS, but I decided to create this Work In Progress issue as a placeholder.

The general idea is to add support for IPFS content identifiers for the feed itself, and content linked in the feed (podcast website, enclosure, etc...).

Maybe this should be a separate namespace, as most of the use cases are not specific to podcasts, but for content distributed through feeds in general.

Is there already any namespace addressing IPFS in feeds? This could branch into a parallel working stream.

**Comment by Jayman2000** [comment]
> Maybe this should be a separate namespace, as most of the use cases are not specific to podcasts, but for content distributed through feeds in general.

I agree. In fact, I don't think that it should be IPFS specific. I'm a big fan of IPFS, but at some point in the future, we'll probably decide that it makes sense to switch from IPFS to something that's even better. Plus, if someone wants to distribute a podcast over FTP, more power to them!

From my perspective, the poblems with using IPFS in an RSS feed are the same as the problems with using any new protocol in an RSS feed:
1. [According to the RSS spec,](https://www.rssboard.org/rss-specification#ltenclosuregtSubelementOfLtitemgt) an enclosure's URL must be an HTTP URL.
1. HTTP(S) fallbacks are needed to allow feeds that use the new protocol to work with existing apps.

For enclosures, I found [Enhanced "enclosures" support in RSS and ATOM Syndication](http://www.crocodile.org/lord/RSSenclosures.pdf). I haven't fully read that document yet, but it seems like a feed could have a standard enclosure with an HTTP URL and an enhanced enclosure with an [IPFS URL](https://docs.ipfs.io/how-to/address-ipfs-on-web/#native-urls) (or any type of URL for that matter).

For the rest of the standard RSS elements, [IPFS URLs can already be used](https://www.rssboard.org/rss-specification#comments), but this would leave users without an HTTP(S) fallback.

At the moment, the Podcast namespace elements reqire HTTPS. One possiblity is to use [IPFS HTTP gateways](https://docs.ipfs.io/how-to/address-ipfs-on-web/#http-gateways) (this could also be done for the standard RSS elements). Podcast apps that don't support IPFS would see them as normal HTTP(S) URLs, and podcast apps that do support IPFS could automatically do a protocol upgrade.

**Comment by theDanielJLewis** [comment]
The RSS spec is from 2001. I think HTTPS was considered only for banks or sites that took payments back then. So shortsighted, yes. But I think there have been no problems using HTTPS enclosures.

**Comment by daveajones** [comment]
@Jayman2000 Thank you for researching prior art, and for the thorough comment.  Very helpful.

We should get some IPFS devs in here if we can.

**Comment by daveajones** [comment]
@theDanielJLewis Agreed.  HTTPS wasn't even formalized in RFC form until 2000.  It was just an afterthought back then.

**Comment by swschilke** [comment]
Isn't that something like a p2p protocoll?

**Comment by vv01f** [comment]
a kind of, it's distributed storage. I still think a URI shouldn't be limited to a protocol at all.

**Comment by swschilke** [comment]
I could agree with this, as long as it makes sense and you can reach the content.

**Comment by vv01f** [comment]
if there can be multiple storages (maybe with different protocols) that might help. just not sure where to get that started.

as mentioned on image, the picture element allows for different files to be added so it can fit the device better. this would be a comparable idea to make file locations more diverse. there also is the URN to refer to a piece without enforcing the location. and magnet links have shown that kind of power for torrents.

**Comment by beaudryj** [comment]
Was thinking about IPFS on my drive today and podcast 2.0... Glad to see there is a thread.. 

I was thinking about this similar to you @vv01f 

why not work on adding a 

<rss xmlns-ipfs:podcast...> and work on helping Podcast 2.0 Developers build their apps to check for a feed on http.. if NULL then parse for an IPFS source? 

```
<?xml version="1.0" encoding="UTF-8"?>
<rss xmlns:podcast="https://github.com/Podcastindex-org/podcast-namespace/blob/main/docs/1.0.md" xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd" version="2.0">
    <channel>
        <title>Podcasting 2.0 Namespace Example</title>
        <description>This is a fake show that exists only as an example of the "podcast" namespace tag usage.</description>
        <link>http://example.com/podcast</link>
        <docs>http://blogs.law.harvard.edu/tech/rss</docs>
        <language>en-US</language>
        <generator>Freedom Controller</generator>
        <pubDate>Fri, 09 Oct 2020 04:30:38 GMT</pubDate>
        <lastBuildDate>Fri, 09 Oct 2020 04:30:38 GMT</lastBuildDate>
        <managingEditor>john@example.com (John Doe)</managingEditor>
        <webMaster>support@example.com (Tech Support)</webMaster>
        <image>
            <url>https://example.com/images/pci_avatar-massive.jpg</url>
            <title>Podcast Feed Template</title>
            <link>https://example.com/show</link>
        </image>

        <podcast:locked owner="podcastowner@example.com">yes</podcast:locked>
        <podcast:funding url="https://example.com/donate">Support the show!</podcast:funding>

        <podcast:location geo="geo:30.2672,97.7431" osm="R113314">Austin, TX</podcast:location>

```

**Comment by Marzal** [comment]
Hi, as per https://github.com/Podcastindex-org/podcast-namespace/discussions/749 and https://github.com/Podcastindex-org/podcast-namespace/discussions/695 we are closing old issues.

IPFS it's available with the `podcast:alternateEnclosure`
https://github.com/search?q=repo%3APodcastindex-org%2Fpodcast-namespace%20IPFS&type=code

Please open a new [Discussions](https://github.com/Podcastindex-org/podcast-namespace/discussions) if you think that the matter it's worth pursuing.
